cloudinit/distros/__init__.py

```diff
@@ -720,6 +725,104 @@ def create_group(self, name, members=None):
                 LOG.info("Added user '%s' to group '%s'", member, name)
 
 
+def _apply_hostname_transformations_to_url(
+    url: str, transformations: "List[Callable[[str], Optional[str]]]"
+) -> "Optional[str]":
+    """
+    Apply transformations to a URL's hostname, return transformed URL.
+
+    This is a separate function because unwrapping and rewrapping only the
+    hostname portion of a URL is complex.
+
+    :param url:
+        The URL to operate on.
+    :param transformations:
+        A list of ``(str) -> Optional[str]`` functions, which will be applied
+        in order to the hostname portion of the URL.  If any function
+        (regardless of ordering) returns None, ``url`` will be returned without
+        any modification.
+
+    :return:
+        A string whose value is ``url`` with the hostname ``transformations``
+        applied, or ``None`` if ``url`` is unparseable.
+    """
+    try:
+        parts = urllib.parse.urlsplit(url)
+    except ValueError:
+        # If we can't even parse the URL, we shouldn't use it for anything
+        return None
+    new_hostname = parts.hostname
+
+    for transformation in transformations:
+        new_hostname = transformation(new_hostname)
+        if new_hostname is None:
+            # If a transformation returns None, that indicates we should abort
+            # processing and return `url` unmodified
+            return url
+
+    new_netloc = new_hostname
+    if parts.port is not None:
+        new_netloc = "{}:{}".format(new_netloc, parts.port)
+    return urllib.parse.urlunsplit(parts._replace(netloc=new_netloc))
+
+
+def _sanitise_mirror_url(url: str) -> "Optional[str]":
```

Comment:
Does this transformation only work on mirror URLs, or can it sanitize any URL?  Should this be in url_helper.py as well?

Reply:
As currently written, it could work for any URL, I believe. That said, my comments in https://github.com/canonical/cloud-init/pull/291/files#r400439439 apply here too: it would need its own testing (though this would be easier to produce using the tests I've already written), and I'm not sure we'll need it generically yet.

Reply:
OK.